demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -58,51 +90,150 @@ AzureIoTResult_t AzureIoTPlatform_WriteBlock( AzureADUImage_t * const pxAduImage
+    // FLASH_TYPEPROGRAM_FAST_AND_LAST
+    if (pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize) {
+        if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_FAST_AND_LAST, (uint32_t)nextWriteAddr, (uint32_t)nextReadAddr) != HAL_OK)
+        {
+            /* Error occurred while writing data in Flash memory*/
+            HAL_FLASH_Lock();
+            return eAzureIoTErrorFailed;
+        }
+    }
+    else { // write last section of blocks 1...n-1 normally
+        if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_FAST, (uint32_t)nextWriteAddr, (uint32_t)nextReadAddr) != HAL_OK)
+        {
+            /* Error occurred while writing data in Flash memory*/
+            HAL_FLASH_Lock();
+            return eAzureIoTErrorFailed;
+        }
+    }
+    HAL_FLASH_Lock();
     return eAzureIoTSuccess;
 }
 
 AzureIoTResult_t AzureIoTPlatform_VerifyImage( AzureADUImage_t * const pxAduImage,
                                                uint8_t * pucSHA256Hash,
                                                uint32_t ulSHA256HashLength )
 {
-    (void)pxAduImage;
-    (void)pucSHA256Hash;
-    (void)ulSHA256HashLength;
+    int xResult;
+    uint32_t ulOutputSize;
+    uint32_t ulReadSize;
 
-    // TODO: Fill in to verify bytes written to flash bank match the SHA256 given by pucSHA256Hash.
+    AZLogInfo( ( "Base64 Encoded Hash from ADU: %.*s", ulSHA256HashLength, pucSHA256Hash ) );
+    xResult = prvBase64Decode( pucSHA256Hash, ulSHA256HashLength, ucDecodedManifestHash, azureiotflashSHA_256_SIZE, ( size_t * ) &ulOutputSize );
 
-    LogInfo( ( "AzureIoTPlatform_VerifyImage()\r\n" ) );
+    if( xResult != eAzureIoTSuccess )
+    {
+        AZLogError( ( "Unable to decode base64 SHA256\r\n" ) );
+        return eAzureIoTErrorFailed;
+    }
 
-    return eAzureIoTSuccess;
-}
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
 
-AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
-{
-    (void)pxAduImage;
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+
+    AZLogInfo( ( "Starting the mbedtls calculation: image size %d\r\n", pxAduImage->ulImageFileSize ) );
 
-    // TODO: Fill in to program board to decide which memory bank to use on reboot.
+    for( size_t ulOffset = 0; ulOffset < pxAduImage->ulImageFileSize; ulOffset += sizeof( ucPartitionReadBuffer ) )
+    {
+        ulReadSize = pxAduImage->ulImageFileSize - ulOffset < sizeof( ucPartitionReadBuffer ) ? pxAduImage->ulImageFileSize - ulOffset : sizeof( ucPartitionReadBuffer );
 
-    LogInfo( ( "AzureIoTPlatform_EnableImage()\r\n" ) );
+        memcpy(ucPartitionReadBuffer, (pxAduImage->xUpdatePartition + ulOffset), ulReadSize);
 
+        mbedtls_md_update( &ctx, ( const unsigned char * ) ucPartitionReadBuffer, ulReadSize );
+
+        if( ( ulOffset % 65536 == 0 ) && ( ulOffset != 0 ) )
+        {
+            printf( "." );
```

Comment:
why printf and not AZLogInfo?

Reply:
This is what we had in the esp sample - I think it's so the `.`s all print to the same line?

Reply:
Yea it was originally because the AzLog stuff will add a newline and the prefix stuff. We can revisit if this is a good idea on small devices like this. Otherwise can maybe use a `putchar()` or something.

Reply:
Going to leave this as an item to address separately for all the samples